### PR TITLE
Geometry_Engine: Optimisation and OutlinesFromLines

### LINKED ISSUE #3259: Geometry_Engine: Boolean operations on line collections are extremely slow
#### Description:

Recently I tried to use Boolean operations on lines, which turned to be _very_ slow, with execution time growing exponentially (or close to) for `BooleanUnion`. By the looks at the code, it is pretty overcooked and could be largely improved by working on parameters of collinear lines. Since (from what I remember) it was me who wrote the original code, happy to have a look myself 😃 TBC

### LINKED ISSUE #3260: Geometry_Engine: add a method joining unordered lines into closed loops
#### Description:

A feature requested by one of the users, to take a set of unordered, collinear lines and turn them into a set of closed loops that can be extracted, e.g. from the following set of lines:
image

... we would get following 6 outlines returned:
image

...this will also enable reconstruction of meshes from their edges.

### LINKED ISSUE #3261: FitLine returns zero line for perfectly aligned points
#### Description:

Numerical glitch, when the input set of points in `FitLine` method is perfectly collinear, the method returns zero length line. Can easily be tested with points `(1,0,0)`, `(2,0,0)` and `(3,0,0)`.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #3259
Closes #3261

### Test files

On [SharePoint](https://burohappold.sharepoint.com/:f:/s/BHoM/EjVR21wEhrhFiuP9CInYFwMBSeeCoSLCJkT6Ci5DubY5Ag?e=2OYTo6).

### Additional comments

Apologies for the massive brain dump and a few different topics covered, but this is all stuff I kept on tweaking while working on various things - happy to split it up into smaller PRs if requested by @peterjamesnugent (volunteered to review) or anyone else eager to have a look.

### CODE REVIEW

**Comment by pawelbaran** [comment]
@BHoMBot check unit-tests

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling
@BHoMBot check core
@BHoMBot check serialisation

**Comment by pawelbaran** [comment]
@BHoMBot check versioning
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

File: Geometry_Engine/Compute/BooleanIntersection.cs [hunk 1]
```diff
@@ -111,15 +90,38 @@ public static Line BooleanIntersection(this Line line, Line refLine, double tole
...
-                return result;
+            return new List<Line> { line }.BooleanIntersection(refLines, tolerance);
+        }
+
+        /***************************************************/
+
 
-            foreach (Line l in refLines.BooleanUnion(tolerance))
+        [Description("Computes the boolean intersection of two collections of lines, e.g. all overlaps of the first set of lines with the reference lines and returns a new collection of lines matching the overlap.")]
+        [Input("lines", "First line to intersect.")]
```
Comment (peterjamesnugent) [inline Geometry_Engine/Compute/BooleanIntersection.cs:100]:
```suggestion
        [Input("lines", "First set of lines to intersect.")]
```

File: Geometry_Engine/Compute/BooleanIntersection.cs [hunk 2]
```diff
@@ -111,15 +90,38 @@ public static Line BooleanIntersection(this Line line, Line refLine, double tole
...
+        [Input("refLines", "A collection of reference lines to intersect with the first line.")]
```
Comment (peterjamesnugent) [inline Geometry_Engine/Compute/BooleanIntersection.cs:101]:
```suggestion
        [Input("refLines", "A set of reference lines to intersect with the first set of lines.")]
```

File: Geometry_Engine/Compute/BooleanIntersection.cs [hunk 3]
```diff
@@ -111,15 +90,38 @@ public static Line BooleanIntersection(this Line line, Line refLine, double tole
...
+        [Input("tolerance", "Tolerance to be used in the method.", typeof(Length))]
+        [Output("intersection", "The collection of lines corresponding to the overlaps of the first line and any of the reference lines.")]
```
Comment (peterjamesnugent) [inline Geometry_Engine/Compute/BooleanIntersection.cs:103]:
```suggestion
        [Output("intersection", "The set of lines corresponding to the overlaps of the first line and any of the reference lines.")]
```

File: Geometry_Engine/Compute/FitLine.cs [hunk 4]
```diff
@@ -35,82 +35,99 @@ public static partial class Compute
 
         public static Line FitLine(this IEnumerable<Point> points, double tolerance = Tolerance.Distance)
```
Comment (peterjamesnugent) [inline Geometry_Engine/Compute/FitLine.cs:36]: worth adding a description here where we can explain the methodologies used, otherwise we are relying on users to look at the code (which I doubt many do).

**Review by peterjamesnugent — COMMENTED** [review]
Happy with changes to `FitLine`, test script works great and tested with the script I added to the issue too. 

Happy with `Colinear` and `Coplanar`, tested using the script.

Happy with `RoundCoordinates`, `BooleanDifference`, `BooleanIntersection` and `BooleanUnion` working as intended, tested using the script.

The unit tests will also demonstrate that the methods have not changed (just extended functionality and dealing with edge cases).

One question I had was on tolerance. 
- For `IsColplanar` the method is true when the difference is 6e-6, but tolerance input is only 1e-6;
- For `IsColinear` the method is true when the different is 3e-6 but tolerance input is only 1e-6;

Are we stacking the tolerances as we feed them through methods and should we be dividing them?

I think we should be consistent with how we refer to `List<T>`, in some places you refer to it in the description as a `set` and in other places its a `collection`. I think I prefer `set` or `list` - but happy for others to chime in.

**Comment by pawelbaran** [comment]
Thanks @peterjamesnugent for the review. I made the tweaks recommended in the code comments.

Regarding the tolerance issues - it is a much bigger problem that boils down to the fact that line fitting, plane fitting, collinearity and coplanarity checks are numerical algorithms rather than geometrical, i.e. the geometrical tolerance cannot be applied 1:1. Long time ago I have spent some time trying to crack the problem, outcome being `Compute.REFTolerance`, but it is still pretty far from perfect.

Happy to bring back the discussion if you think the current precision is not good enough. But it is definitely beyond the scope of this PR 😉

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

**Review by peterjamesnugent — APPROVED** [review]
Happy that the last commit was all descriptions and making use of the `IsNull`, so my previous review of functionality is still valid. 

I assume there will be a seperate PR to update unit tests to include (at least some) of the cases in the test scripts.

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: unit-tests

**Comment by peterjamesnugent** [comment]
@BHoMBot check ready-to-merge
